### PR TITLE
Generate ConfigProto from config nodes

### PR DESCRIPTION
**Pull request by eoogbe**
Revised gapic config advisor:

1. Replace configgen transformers with config nodes
2. **Generate ConfigProto from config nodes**
3. Replace discogapic configgen transformers with config nodes
4. Replace transformer-based adviser with config nodes
5. Allow refresh in configgen
6. Cleanup unused classes
7. Add advising rules

`ConfigNode`s are read from the `ConfigYamlReader`, then transformed to a `ConfigProto` by the `MessageGenerator`. Optionally, the mergers can add missing fields using the `MissingFieldTransformer` before running through the `MessageGenerator`.

### CODE REVIEW

File: src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java [hunk 1]
```diff
@@ -121,14 +123,9 @@ public DiscoGapicGeneratorApi(ToolOptions options) {
...
-    ConfigSource configSource = loadConfigFromFiles(configFileNames);
-    if (configSource == null) {
-      throw new IOException("Failed to load config source.");
-    }
-
-    ConfigProto configProto = (ConfigProto) configSource.getConfig();
-    if (configProto == null) {
-      throw new IOException("Failed to cast config proto.");
+    ConfigProto configProto = loadConfigFromFiles(configFileNames);
+    if (ConfigProto.getDefaultInstance().equals(configProto)) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/DiscoGapicGeneratorApi.java]: As above: do we need a null check that prints out collected diag errors?

File: src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import java.util.List;
+import org.yaml.snakeyaml.nodes.MappingNode;
+import org.yaml.snakeyaml.nodes.Node;
+import org.yaml.snakeyaml.nodes.NodeTuple;
+import org.yaml.snakeyaml.nodes.ScalarNode;
+import org.yaml.snakeyaml.nodes.SequenceNode;
+
+public class ConfigYamlNodeReader {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java]: Class-level comments, please

File: src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+  private final List<String> lines;
+  private final ConfigHelper helper;
+
+  public ConfigYamlNodeReader(List<String> lines, ConfigHelper helper) {
+    this.lines = lines;
+    this.helper = helper;
+  }
+
+  public ConfigNode readMessageNode(int prevLine, MappingNode node, Descriptor messageType) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java]: Comments please. (What do the `node` and `messageType` represent? What is contained in the `ConfigNode` that they're read into?)

File: src/main/java/com/google/api/codegen/configgen/InitialConfigLocationGenerator.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package com.google.api.codegen.configgen;
+
+import com.google.api.tools.framework.model.Location;
+import com.google.api.tools.framework.model.SimpleLocation;
+
+/** Implements LocationGenerator for an initial config. */
+public class InitialConfigLocationGenerator implements LocationGenerator {
+  @Override
+  public Location getLocation() {
+    return new SimpleLocation("gapic config");
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/InitialConfigLocationGenerator.java]: Is it possible to use the output file name (e.g., `pubsub_gapic.yaml`)?

File: src/main/java/com/google/api/codegen/configgen/MessageGenerator.java [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+import com.google.api.codegen.configgen.nodes.ListItemConfigNode;
+import com.google.api.codegen.configgen.nodes.ScalarConfigNode;
+import com.google.api.tools.framework.util.VisitsBefore;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Message;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Generates a Message from a ConfigNode representation. */
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/MessageGenerator.java]: nit: "Message" is ambiguous: "protocol buffer Message"?

File: src/main/java/com/google/api/codegen/configgen/MessageGenerator.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    this(messageBuilder, null);
+  }
+
+  private MessageGenerator(Message.Builder messageBuilder, FieldDescriptor field) {
+    this.messageBuilder = messageBuilder;
+    this.field = field;
+  }
+
+  @VisitsBefore
+  boolean generate(FieldConfigNode node) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/MessageGenerator.java]: optional: seems a little strange for a public class to have mostly package-private methods

File: src/main/java/com/google/api/codegen/configgen/MissingFieldTransformer.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import com.google.api.codegen.configgen.nodes.FieldConfigNode;
+
+/** Generates a FieldConfigNode if missing from its parent. */
+public class MissingFieldTransformer {
+  private final String name;
+  private final ConfigNode parent;
+  private final ConfigNode prev;
+
+  /**
+   * Creates a MissingFieldTransformer that prepends a FieldConfigNode to the given parent if one
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/MissingFieldTransformer.java]: I'm having trouble parsing this. Could you give an example?

File: src/main/java/com/google/api/codegen/configgen/MessageGenerator.java [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+import com.google.api.codegen.configgen.nodes.ScalarConfigNode;
+import com.google.api.tools.framework.util.VisitsBefore;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Message;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Generates a Message from a ConfigNode representation. */
+public class MessageGenerator extends NodeVisitor {
```
[the remark this answers is not in the document]
Comment (eoogbe) [reply src/main/java/com/google/api/codegen/configgen/MessageGenerator.java]: That's how `GenericVisitor` specifies the implementation of visitors. Those default methods are what make the public method `visit` work.

File: src/main/java/com/google/api/codegen/configgen/NodeFinder.java [hunk 9]
```diff
@@ -15,12 +15,45 @@
...
+    for (ConfigNode childNode : getChildren(parentNode)) {
+      if (value.equals(childNode.getText())) {
+        return childNode;
+      }
+    }
+
+    return new NullConfigNode();
+  }
+
+  public static boolean hasChild(ConfigNode parentNode) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/configgen/NodeFinder.java]: nit: Comment this method, and maybe consider changing the title. From a readability perspective, it's hard to tell when browsing the code how `hasChild(node)` should be different from `node.getChild().isPresent()`. What does "having a child" mean that requires so many checks?

File: src/main/java/com/google/api/codegen/CodeGeneratorApi.java [hunk 10]
```diff
@@ -195,15 +190,24 @@ public void error(String message, Object... args) {
...
     if (model.getDiagCollector().getErrorCount() > 0) {
       return null;
     }
-    ImmutableMap<String, Message> supportedConfigTypes =
-        ImmutableMap.<String, Message>of(
-            ConfigProto.getDescriptor().getFullName(), ConfigProto.getDefaultInstance());
-    return MultiYamlReader.read(model.getDiagCollector(), configFiles, supportedConfigTypes);
+
+    ProtoConfigMerger configMerger = new ProtoConfigMerger();
+    MessageGenerator messageGenerator = new MessageGenerator(ConfigProto.newBuilder());
```
[the remark this answers is not in the document]
Comment (eoogbe) [reply src/main/java/com/google/api/codegen/CodeGeneratorApi.java]: The `MessageGenerator` is reused within itself recursively to generate the inner messages of the `ConfigProto`.

File: src/main/java/com/google/api/codegen/configgen/mergers/ConfigMerger.java [hunk 11]
```diff
@@ -16,16 +16,23 @@
...
+import com.google.api.codegen.configgen.MissingFieldTransformer;
+import com.google.api.codegen.configgen.NodeFinder;
 import com.google.api.codegen.configgen.nodes.ConfigNode;
 import com.google.api.codegen.configgen.nodes.FieldConfigNode;
+import com.google.api.codegen.configgen.nodes.ScalarConfigNode;
 import com.google.api.codegen.configgen.nodes.metadata.DefaultComment;
 import com.google.api.codegen.configgen.nodes.metadata.FixmeComment;
+import java.io.File;
 
 /** Merges the gapic config from an ApiModel into a ConfigNode. */
```
Comment (ethanbao) [inline src/main/java/com/google/api/codegen/configgen/mergers/ConfigMerger.java]: Not part of this PR. But could you clarify the merging logic here? For example, is overriding allowed under current implementation?

Comment (eoogbe) [reply src/main/java/com/google/api/codegen/configgen/mergers/ConfigMerger.java]: The merging for this pr checks if the default fields are present and adds them if not. The manual edits can add fields, change fields, but not remove default fields. I think we will need some sort of stateful merging to preserve deletions.

File: src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+  private ConfigNode readSingularField(int prevLine, Node node, FieldDescriptor field) {
+    if (field.getType() == FieldDescriptor.Type.MESSAGE) {
+      return readMessageNode(prevLine, node, field);
+    }
+
+    return readScalarNode(node, field);
+  }
+
+  private ConfigNode readCommentNode(int prevLine, Node node, ConfigNode configNode) {
```
[the remark this answers is not in the document]
Comment (eoogbe) [reply src/main/java/com/google/api/codegen/configgen/ConfigYamlNodeReader.java]: Done
